### DISCUSSION TITLE
Add devpi_getp for optional password provisioning

### DISCUSSION THREAD

**Discussion by johan-westin-private** [devpi/devpi#943] [Ideas]
Hi,

Inspired by restic, it would be nice to have devpi_getp support to provide password.

Naive implementation (in login.py):

```
@hookimpl(trylast=True)
def devpiclient_get_password(url, username):
    def _fallback(_url, _username):
        if sys.version_info < (3,):
            # getpass.getpass() does not support Unicode prompts on Python 2
            _username = _username.encode(sys.stdout.encoding)
            _url = _url.encode(sys.stdout.encoding)

        return getpass.getpass("password for user %s at %s: " % (_username, _url))

    maybe_cmd = shutil.which(os.environ.get('DEVPI_PASSWORD_COMMAND', 'devpi_getp'))
    if maybe_cmd:
        p = Popen(maybe_cmd, stderr=PIPE, stdout=PIPE)
        out, err = p.communicate()
        passwd = out.decode()
        if p.returncode == 0 and passwd:
            return passwd
        else:
            return _fallback(url, username)
    else:
        return _fallback(url, username)


```

**Comment by fschulze** [comment]
Hi!

I tried to understand your intend, but am at a loss. What is ``devpi_getp`` and what is the goal of your proposal? What is your use-case?

**Comment by johan-westin-private** [comment]
Good morning,

It is a way to "autoload" the password, so you don't need to write it in 
the console.

devpi_getp is the default executable name that provides the password, how 
this is done is unknown for the devpi client.

A possible input to devpi_getp could be the username.

Look at 
https://restic.readthedocs.io/en/latest/040_backup.html#environment-variables 
where RESTIC_PASSWORD_COMMAND default is restic_getp.

Happy New Year,
Johan Westin

Den 31 december 2022 07:40:03 skrev Florian Schulze ***@***.***>:

**Comment by fschulze** [comment]
Ok, there may be existing solutions that can solve your use-case. For one there is a devpi-client hook in [``keyring``](https://pypi.org/project/keyring/) and there is the upcoming [devpi-tokens](https://github.com/devpi/devpi-tokens).

**Reply by johan-westin-private** [reply]
Thanks for tips and for devpi* !
